### PR TITLE
wicked: ping_with_timeout should not fail when there is one packet lost

### PR DESCRIPTION
**Pull request by asmorodskyi**
ping_with_timeout should not fail when there is one packet lost

### CODE REVIEW

File: lib/wickedbase.pm [hunk 1]
```diff
@@ -223,7 +223,7 @@ sub ping_with_timeout {
     my ($self, %args) = @_;
     $args{ip_version}         //= 'v4';
     $args{timeout}            //= '60';
-    $args{proceed_on_failure} //= 0;
```
Comment (cfconrad) [inline lib/wickedbase.pm]: To me it looks like we disabling that function, or?

File: lib/wickedbase.pm [hunk 2]
```diff
@@ -235,7 +235,8 @@ sub ping_with_timeout {
             if ($args{count_success} > 1) {
                 my $cnt = $args{count_success} - 1;
                 $ping_command =~ s/\s-c\s1\s+/ -c $cnt /;
-                validate_script_output($ping_command, sub { /\s+0% packet loss/ });
+                my $ping_out = script_output($ping_command, proceed_on_failure => $args{proceed_on_failure});
+                record_info('WARNING PING FAILED', $ping_out) unless ($ping_out =~ /\s+0% packet loss/ )
```
Comment (cfconrad) [inline lib/wickedbase.pm]: Should it be `record_softfailure` ?
I actually would vote for specifying a threshold in `%` of failed packages.

File: lib/wickedbase.pm (hunk 1 above)
Comment (asmorodskyi) [reply lib/wickedbase.pm]: agree , I miss that . PR updated WDYT now ?

File: lib/wickedbase.pm (hunk 2 above)
Comment (asmorodskyi) [reply lib/wickedbase.pm]: 1. there should be no softfailures without bug ID. 
2.  theshold added

File: lib/wickedbase.pm [hunk 3]
```diff
@@ -235,17 +235,17 @@ sub ping_with_timeout {
             if ($args{count_success} > 1) {
                 my $cnt = $args{count_success} - 1;
                 $ping_command =~ s/\s-c\s1\s+/ -c $cnt /;
-                validate_script_output($ping_command, sub { /\s+0% packet loss/ });
+                my $ping_out = script_output($ping_command, proceed_on_failure => $args{ignore_packet_loss});
+                $ping_out =~ /, (\d[1-3])% packet loss/;
```
Comment (cfconrad) [inline lib/wickedbase.pm]:
```suggestion
                $ping_out =~ /, (\d{1,3})% packet loss/;
```

File: lib/wickedbase.pm (hunk 1 above)
Comment (cfconrad) [reply lib/wickedbase.pm]: I only found one user of this method, which use `proceed_on_failure =>1`

```
tests/wicked/basic/sut/t01_basic.pm:60:    die('IP should not be reachable') if ($self->ping_with_timeout(ip => '10.0.2.2', timeout => '2', proceed_on_failure => 1));
```
It looks like, we test there, that the connection to a specific host should _not_ work. I would keep the logic to check for failed connections.

File: lib/wickedbase.pm [hunk 4]
```diff
@@ -235,17 +235,17 @@ sub ping_with_timeout {
...
+                #we treat interface in workable state if it manage to echo more than half packets
+                record_info('WARNING PING FAILED', $ping_out) unless ($1 < 50 )
```
Comment (cfconrad) [inline lib/wickedbase.pm]:
```suggestion
                $args{max_packed_lost} //= 50;
                if ($1 > $args{max_packed_lost} ) {
                         die("PING FAILED: $ping_out") unless($args{proceed_on_failure});
                         record_info('WARNING PING FAILED', $ping_out);
                }
                
```

File: lib/wickedbase.pm [hunk 5]
```diff
@@ -235,17 +235,17 @@ sub ping_with_timeout {
...
             return 1;
         }
         $timeout -= 1;
         sleep 5;
     }
-    if (!$args{proceed_on_failure}) {
-        die('PING failed: ' . $ping_command);
-    }
-    return 0;
+    die('PING failed: ' . $ping_command);
```
Comment (cfconrad) [inline lib/wickedbase.pm]: same as above, I would keep the "proceed_on_failure"

File: lib/wickedbase.pm [hunk 6]
```diff
@@ -233,9 +233,18 @@ sub ping_with_timeout {
     while ($timeout > 0) {
         if (script_run($ping_command) == 0) {
             if ($args{count_success} > 1) {
-                my $cnt = $args{count_success} - 1;
+                my $cnt       = $args{count_success} - 1;
+                my $threshold = 50;
                 $ping_command =~ s/\s-c\s1\s+/ -c $cnt /;
-                validate_script_output($ping_command, sub { /\s+0% packet loss/ });
+                my $ping_out = script_output($ping_command, proceed_on_failure => 1);
```
Comment (jlausuch) [inline lib/wickedbase.pm]: if we force proceed_on_failure => 1 here, we don't need it as  input parameter (line 226).

Comment (asmorodskyi) [reply lib/wickedbase.pm]: we do need it later , see line 245 ( old version ) / 254 ( new version )

File: lib/wickedbase.pm [hunk 7]
```diff
@@ -233,9 +233,18 @@ sub ping_with_timeout {
...
+                $ping_out =~ /, (\d{1,3})% packet loss/;
+                #we treat interface in workable state if it manage to echo more than half packets
```
Comment (jlausuch) [inline lib/wickedbase.pm]: I wouldn't state this without a good reason. Why 50%?  I would vote for having threshold as input parameter and let the user of this function decide

File: lib/wickedbase.pm [hunk 8]
```diff
@@ -233,9 +233,18 @@ sub ping_with_timeout {
...
+                if ($1 > $threshold) {
+                    die('PING EXCEED THRESHOLD ' . $threshold . '\n' . $ping_out);
+                }
+                elsif (!$1) {
```
Comment (cfconrad) [inline lib/wickedbase.pm]: Isn't it `elsif($1)` ?

File: lib/wickedbase.pm [hunk 9]
```diff
@@ -233,9 +233,18 @@ sub ping_with_timeout {
...
+                    record_info('WARNING', 'PING EXCEED THRESHOLD ' . $threshold . '\n' . $ping_out);
```
Comment (cfconrad) [inline lib/wickedbase.pm]:
```suggestion
                    record_info('WARNING', 'PING with packed loss ' . $threshold . '%\n' . $ping_out);
```

File: lib/wickedbase.pm (hunk 8 above)
Comment (asmorodskyi) [reply lib/wickedbase.pm]: yeah , you right

File: lib/wickedbase.pm (hunk 6 above)
Comment (cfconrad) [reply lib/wickedbase.pm]: It's used in https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/7600/files#diff-7c6669e0bb0ed1f0eed4a62dc81f200dR254

File: lib/wickedbase.pm (hunk 7 above)
Comment (asmorodskyi) [reply lib/wickedbase.pm]: fixed

File: lib/wickedbase.pm (hunk 9 above)
Comment (asmorodskyi) [reply lib/wickedbase.pm]: fixed

File: lib/wickedbase.pm (hunk 2 above)
Comment (cfconrad) [reply lib/wickedbase.pm]: ok

File: lib/wickedbase.pm [hunk 10]
```diff
@@ -235,7 +236,15 @@ sub ping_with_timeout {
...
                 $ping_command =~ s/\s-c\s1\s+/ -c $cnt /;
-                validate_script_output($ping_command, sub { /\s+0% packet loss/ });
+                my $ping_out = script_output($ping_command, proceed_on_failure => 1);
+                $ping_out =~ /, (\d{1,3})% packet loss/;
+                #we treat interface in workable state if it manage to echo more than half packets
+                if ($1 > $args{threshold}) {
+                    die('PING EXCEED THRESHOLD ' . $args{threshold} . '%\n' . $ping_out);
+                }
+                elsif (!$1) {
+                    record_info('WARNING', 'PING with packed loss ' . $args{threshold} . '%\n' . $ping_out);
```
Comment (jlausuch) [inline lib/wickedbase.pm]:
```suggestion
                    record_info('WARNING', 'PING with packed loss higher than threshold' . $args{threshold} . '%\n' . $ping_out);
```

Comment (asmorodskyi) [reply lib/wickedbase.pm]: it is lower than threshold actually if it is higher we dying above ...

Comment (jlausuch) [reply lib/wickedbase.pm]: you are right, then, something like `"WARNING", "PING with 25% packet loss. Threshold is 50%. ... "`

File: lib/wickedbase.pm [hunk 11]
```diff
@@ -235,7 +236,15 @@ sub ping_with_timeout {
...
                 $ping_command =~ s/\s-c\s1\s+/ -c $cnt /;
-                validate_script_output($ping_command, sub { /\s+0% packet loss/ });
+                my $ping_out = script_output($ping_command, proceed_on_failure => 1);
+                $ping_out =~ /, (\d{1,3})% packet loss/;
+                #we treat interface in workable state if it manage to echo more than half packets
+                if ($1 > $args{threshold}) {
+                    die('PING EXCEED THRESHOLD ' . $args{threshold} . '%\n' . $ping_out);
+                }
+                elsif ($1) {
+                    record_info('WARNING', sprintf('PING with %d% packet loss. Threshold is %d% \n %s', $1, $args{threshold}, $ping_out));
```
Comment (cfconrad) [inline lib/wickedbase.pm]:
```suggestion
                    record_info('WARNING', sprintf('PING with %d%% packet loss. Threshold is %d%% \n %s', $1, $args{threshold}, $ping_out));
```

Comment (cfconrad) [reply lib/wickedbase.pm]: Otherwise it will produce a warning:

```
perl -w -e  'print sprintf("% ");'
```

Comment (asmorodskyi) [reply lib/wickedbase.pm]: fixed

**Pull request merged**